### PR TITLE
fix issue 19928

### LINKED ISSUE #19928: With the LDAP_AUTH or OIDC_AUTH, what if the current user is a group member and the group is granted as admin role?
With the LDAP_AUTH or OIDC_AUTH, what if the current user is a group member and the group is granted as admin role?

_Originally posted by @stonezdj in https://github.com/goharbor/harbor/pull/19543#discussion_r1385879526_
       
We ran into this issue. A workaround for the user who wants to create a robot account is to add the user to the project directly as ProjectAdmin, which is somehow dirty.

### PR DESCRIPTION
**Pull request by wy65701436**
it needs to consider the user who is in any group that has been granted with the project admin role.

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #19928

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/pkg/project/dao/dao.go [hunk 1]
```diff
@@ -202,19 +203,39 @@ func (d *dao) ListRoles(ctx context.Context, projectID int64, userID int, groupI
...
 	}
 
+	var membersG []models.Member
+	if len(user.GroupIDs) > 0 {
+		var params []interface{}
+		params = append(params, user.GroupIDs)
+		sqlG := fmt.Sprintf(`select b.* from project as a 
+    		left join project_member as b on a.project_id = b.project_id 
+           	where a.deleted = 'f' and b.entity_id in ( %s ) and b.entity_type = 'g' and b.role = 1;`, orm.ParamPlaceholderForIn(len(user.GroupIDs)))
+		_, err = o.Raw(sqlG, params).QueryRows(&membersU)
```
Comment (zyyw) [inline src/pkg/project/dao/dao.go:226]: `_, err = o.Raw(sqlG, params).QueryRows(&membersG)`

**Review by zyyw — APPROVED** [review]

**Review by MinerYang — APPROVED** [review]
lgtm

File: src/server/v2.0/handler/permissions.go [hunk 2]
```diff
@@ -16,11 +16,12 @@ package handler
 
 import (
 	"context"
-
+
```
Comment (MinerYang) [inline src/server/v2.0/handler/permissions.go:19]: heading white space should be removed
 (gogofmt -w src/server/v2.0/handler/permissions.go )

File: src/pkg/project/dao/dao.go [hunk 3]
```diff
@@ -202,19 +203,39 @@ func (d *dao) ListRoles(ctx context.Context, projectID int64, userID int, groupI
...
+    		left join project_member as b on a.project_id = b.project_id 
+           	where a.deleted = 'f' and b.entity_id in ( %s ) and b.entity_type = 'g' and b.role = 1;`, orm.ParamPlaceholderForIn(len(user.GroupIDs)))
+		_, err = o.Raw(sqlG, params).QueryRows(&membersG)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var members []models.Member
+	if len(membersU) > 0 {
```
Comment (stonezdj) [inline src/pkg/project/dao/dao.go:233]: No need to check len(membersU)>0

File: src/server/v2.0/handler/permissions.go [hunk 4]
```diff
@@ -57,15 +58,14 @@ func (p *permissionsAPI) GetPermissions(ctx context.Context, _ permissions.GetPe
...
-		user, err := p.uc.GetByName(ctx, secCtx.GetUsername())
-		if err != nil {
-			return p.SendError(ctx, err)
-		}
-		is, err := p.mc.IsProjectAdmin(ctx, user.UserID)
-		if err != nil {
-			return p.SendError(ctx, err)
+		if sc, ok := secCtx.(*local.SecurityContext); ok {
+			user := sc.User()
+			is, err := p.mc.IsProjectAdmin(ctx, *user)
```
Comment (stonezdj) [inline src/server/v2.0/handler/permissions.go:63]: replace `is` with `isProjectAdmin`

File: src/pkg/project/dao/dao.go [hunk 5]
```diff
@@ -202,19 +203,39 @@ func (d *dao) ListRoles(ctx context.Context, projectID int64, userID int, groupI
...
+		members = append(members, membersU...)
+	}
+	if len(membersG) > 0 {
```
Comment (stonezdj) [inline src/pkg/project/dao/dao.go:236]: Dito

File: src/pkg/project/dao/dao.go (hunk 3 above)
Comment (wy65701436) [reply src/pkg/project/dao/dao.go:233]: let's keep it for readability.

File: src/server/v2.0/handler/permissions.go (hunk 4 above)
Comment (wy65701436) [reply src/server/v2.0/handler/permissions.go]: resolved.

File: src/server/v2.0/handler/permissions.go (hunk 2 above)
Comment (wy65701436) [reply src/server/v2.0/handler/permissions.go]: resolved.

**Pull request merged**